File: programmers/고득점 Kit/DFSBFS/1844.py
# Solved (15m)
from collections import deque
def bfs(maps):
    move = [(-1,0), (1,0), (0,1), (0,-1)]
    queue = deque([(0, 0)])
    visited[0][0] = True
    while queue:
        y, x = queue.popleft()
        if y==len(maps)-1 and x==len(maps[0])-1:
            return visited[y][x]
            
        for dx, dy in move:
            ny = y + dy
            nx = x + dx
            if 0<=ny<len(maps) and 0<=nx<len(maps[0]):
                if not visited[ny][nx] and maps[ny][nx]==1:
                    queue.append((ny, nx))
                    visited[ny][nx] = visited[y][x] + 1
    
    return None
    
def solution(maps):
    global visited
    answer = 0
    visited = [[False]*len(maps[0]) for _ in range(len(maps))]
    result = bfs(maps)
    if result:
        return result
    else:
        return -1

File: programmers/고득점 Kit/DFSBFS/1844.java
import java.util.*;
class Solution {
    static int answer = -1;
    public int solution(int[][] maps) {
        bfs(maps);
        return answer;
    }
    static void bfs(int[][] maps) {
        int n = maps.length;
        int m = maps[0].length;
        Queue<int[]> queue = new LinkedList<>();
        queue.offer(new int[]{0, 0, 1});
        boolean[][] visited = new boolean[n][m];
        visited[0][0] = true;
        int[] dx = {-1, 1, 0, 0};
        int[] dy = {0, 0, -1, 1};
        while(!queue.isEmpty()) {
            int[] tmp = queue.poll();
            int x = tmp[0];
            int y = tmp[1];
            int cnt = tmp[2];
            if(x==m-1 && y==n-1) {
                answer = cnt;
                return;
            }
            for(int i=0; i<4; i++) {
                int nx = x + dx[i];
                int ny = y + dy[i];
                if(ny>=0&&ny<n&&nx>=0&&nx<m) {
                    if(!visited[ny][nx] && maps[ny][nx]==1) {
                        queue.offer(new int[]{nx, ny, cnt+1});
                        visited[ny][nx] = true;
                    }
                }
            }
        }
        return;
    }
}